mkosi/kmod.py

```diff
@@ -55,6 +55,39 @@ def filter_kernel_modules(
     return sorted(modules)
 
 
+def filter_firmware(
+    root: Path,
+    firmware: set[Path],
+    *,
+    include: Iterable[str],
+    exclude: Iterable[str],
+) -> list[Path]:
+    firmwared = Path("usr/lib/firmware")
+    with chdir(root):
+        all_firmware = set(firmwared.rglob("*"))
+
+    keep = set()
+    if include:
+        regex = re.compile("|".join(include))
+        for f in all_firmware:
+            rel = os.fspath(Path(*f.parts[3:]))
+            if regex.search(rel):
+                keep.add(rel)
+                firmware.add(f)
+
+    if exclude:
+        remove = set()
+        regex = re.compile("|".join(exclude))
+        for f in firmware:
+            rel = os.fspath(Path(*f.parts[3:]))
+            if rel not in keep and regex.search(rel):
+                remove.add(f)
+
+        firmware -= remove
```

Comment:
I would switch the operations around, first do the `exclude` logic, and then the `include` logic. It doesn't matter too much if we remove stuff first that we'll add back later since insert/remove from a set is a constant time operation anyway.

Switching this around allows you to get rid of the `keep` variable.